Enable filtering for tasks and custom rewards

JiriChara opened the pull request:
Add search input box next to the tags. When user changes the value of
the input box then it's propagated to the user service as filterQuery
string. filterQuery string is then used in the tasks template to filter
tasks by text.

JiriChara committed:
Enable filtering for tasks and custom rewards

Add search input box next to the tags. When user changes the value of
the input box then it's propagated to the user service as filterQuery
string. filterQuery string is then used in the tasks template to filter
tasks by text.

JiriChara committed:
Remove unused dependency

Alys commented:
Fast work! And you included a test! We like that. :)

Could you please add a screenshot of the tasks page with the search input box in it? It doesn't have to be the whole page, just enough to get a good feel for how the box will look with the other elements around it.

Also, a useful tip for github pull requests: If you enter:
 `fixes https://github.com/HabitRPG/habitrpg/issues/5385`
(i.e., the URL for an issue), then github creates a link between the two and it also automatically closes the issue when the PR is merged. Like so:  fixes https://github.com/HabitRPG/habitrpg/issues/5385 :)

JiriChara commented:
Hi Alys,

thank you for the advice! See the screenshots below:

![2015-06-11-100344_469x450_scrot](https://cloud.githubusercontent.com/assets/668763/8102549/3c2ac57a-1021-11e5-879d-c694111543f3.png)

mouseover:
![2015-06-11-100327_739x441_scrot](https://cloud.githubusercontent.com/assets/668763/8102547/3c28575e-1021-11e5-995f-da5ff1d34864.png)

focus:
![2015-06-11-100312_702x374_scrot](https://cloud.githubusercontent.com/assets/668763/8102550/3c2cdedc-1021-11e5-9564-28e52bfe9892.png)

no focus/no mouseover
![2015-06-11-100209_620x219_scrot](https://cloud.githubusercontent.com/assets/668763/8102548/3c291356-1021-11e5-8be1-c947b982e1bc.png)

Alys commented:
@lemoness @AnitaYCheng what do you think of the UI for this?

JiriChara commented:
I will do little bit of cosmetics, because the search field is not wrapped very nicely on small screens. I also don't like the placeholder text ("Search"). Do you have any idea on how to rename it?

crookedneighbor commented on website/views/shared/tasks/task.jade:
```diff
@@ -1,4 +1,4 @@
-li(bindonce='list', id='task-{{::task.id}}', ng-repeat='task in obj[list.type+"s"] | conditionalOrderBy: list.view=="dated":"date"', class='task {{Shared.taskClasses(task, user.filters, user.preferences.dayStart, user.lastCron, list.showCompleted, main)}}', ng-click='spell && (list.type != "reward") && castEnd(task, "task", $event)', ng-class='{"cast-target":spell && (list.type != "reward"), "loc ...
+li(bindonce='list', id='task-{{::task.id}}', ng-repeat='task in obj[list.type+"s"] | filter:{text: obj.filterQuery} | conditionalOrderBy: list.view=="dated":"date"', class='task {{Shared.taskClasses(task, user.filters, user.preferences.dayStart, user.lastCron, list.showCompleted, main)}}', ng-click='spell && (list.type != "reward") && castEnd(task, "task", $event)', ng-class='{"cast-target":spell ...
```
Is there a reason not to filter by notes as well?

JiriChara replied:
That's a good point! I will fix it as soon I'll have time.

crookedneighbor commented on website/public/js/controllers/filtersCtrl.js:
```diff
@@ -30,6 +31,12 @@ habitrpg.controller("FiltersCtrl", ['$scope', '$rootScope', 'User', 'Shared',
       // User.save();
     };
 
+    $scope.$watch('filterQuery', function (newValue, oldValue) {
```
Do we really need a watch to accomplish this?

JiriChara replied:
We can assign user to the scope of the controller and then use it's `filterQuery` attribute as a `ng-model` for the search field. I thought that a watcher might be more transparent for the programmers - so they can directly see the relation between user service and filter controller w/o digging into the view and it is easier to test too. I mean, I can still change it if you like!

crookedneighbor replied:
Not sure. Still thinking about it. 

In general, we try to limit the number of watches going on. We already have so many :disappointed:

JiriChara committed:
Make sure search filter has same margin as tags

crookedneighbor commented on website/views/main/filters.jade:
```diff
@@ -1,6 +1,7 @@
 .container-fluid
   .row
     .filters(ng-controller='FiltersCtrl')
+      input.filters-search(type='text', placeholder=env.t('search'), ng-model='filterQuery')
```
How would this look using bootstrap's input styling instead?

JiriChara committed:
Implement filter tasks by text & notes

In order to be able to filter tasks by it's text or notes - we need an
extra filter to be implemented, because angular does not support OR
condition for it's default filter directive.

JiriChara committed:
Change the styling of filter box to bootstrap

JiriChara commented:
The new bootstrap design:

![2015-06-11-204512_471x98_scrot](https://cloud.githubusercontent.com/assets/668763/8115393/47700d46-107b-11e5-90bd-d6351d115bd8.png)

lemoness commented:
Bootstrap looks better, but it seems a little tall.

I'm also not sold on the placement - it seems to imply to me that you use
it to search for tags.Is there a way to have it on the right side of the
page instead?

On Thu, Jun 11, 2015 at 11:51 AM, Jiří Chára notifications@github.com
wrote:

> The new bootstrap design:
> 
> [image: 2015-06-11-204512_471x98_scrot]
> https://cloud.githubusercontent.com/assets/668763/8115393/47700d46-107b-11e5-90bd-d6351d115bd8.png
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5389#issuecomment-111237840.

JiriChara commented:
The height of the search box is same as the height of those tags. I will put the search field to the right

JiriChara committed:
Put tasks filter field to the right

JiriChara commented:
@lemoness what do you think about the new position and styling?

![2015-06-11-221756_1057x174_scrot](https://cloud.githubusercontent.com/assets/668763/8117342/be797114-1087-11e5-9e84-729703468f59.png)

lemoness commented:
Looks great to me!

On Thu, Jun 11, 2015 at 1:19 PM, Jiří Chára notifications@github.com
wrote:

> @lemoness https://github.com/lemoness what do you think about the new
> position and styling?
> 
> [image: 2015-06-11-221756_1057x174_scrot]
> https://cloud.githubusercontent.com/assets/668763/8117342/be797114-1087-11e5-9e84-729703468f59.png
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5389#issuecomment-111264546.

Alys commented:
I like that a lot! It looks very clean and neat.

However what does it look like when there's a lot of tags? I've seen people with three rows of tags. Do they wrap nicely around the search box with a bit of horizontal space to separate them?

JiriChara commented:
![2015-06-11-223230_923x140_scrot](https://cloud.githubusercontent.com/assets/668763/8117624/d3dafa80-1089-11e5-96b5-4bed058094a1.png)

![2015-06-11-223201_402x230_scrot](https://cloud.githubusercontent.com/assets/668763/8117625/d3f38ca8-1089-11e5-8e99-288f39d7fc03.png)

Alys commented:
I like that. In the first shot, the blank space between the tag controls and the search box could be considered by some to be a waste of space, but I think it looks nice, and uncluttered. The performance on a small screen is great (thanks for thinking to include that).

crookedneighbor committed:
Adjust PR #5389

* Convert $watch to ng-change and re-wrote test
* Removed unnecessary styles
* Use .form-control instead of class='form-control'

crookedneighbor committed:
Merge branch 'develop' into blade-feature/search

JiriChara committed:
Merge pull request #1 from crookedneighbor/blade-feature/search

Blade feature/search

crookedneighbor commented:
This will count toward a contributor tier. What's your UUID? It can be found in Settings > API > User Id

JiriChara commented:
@crookedneighbor cool thanks, my UID is: a85f4815-af74-482e-9d26-5fd50fb071c9 :+1:

Alys commented:
I've just discovered a problem with this, and unfortunately the revert option no longer works. It's not super critical though, and just for the record, @JiriChara I LOVE this feature and you still get contributor credit for it. :)

If you have a search query in place and you go to another page in the website, then back to the tasks page, the search query is still in place but the search box is empty (as if there is no current search). This will undoubtably lead to people reporting "bugs" that their tasks have disappeared or been deleted.

The way to recover from it is to type anything into the search box and then delete that text, but I think we need a better fix, and also an obvious indication that a search filter has been applied. 

We should probably also clear the search filter when moving away from (or back to) the tasks page. Even if we do that, I think it would help if there was also an indicator that a filter was in place. Perhaps text just above the task columns that says "**showing only tasks containing [search term goes here]**". There could be a "clear filter" button next to that text. (We could also make that text describe any tags that are selected, which would help to reduce the bug reports from tasks being hidden by forgotten tag filters.)

Any other ideas?

Are we planning to do a deployment to production in the next day or two? If so, do we want to revert this manually before then, or do we think we can put in any fix or workaround soon enough that it won't matter?

crookedneighbor commented:
Ohh, good catch. I know how to fix it.

crookedneighbor commented:
@Alys this should fix it: https://github.com/HabitRPG/habitrpg/commit/b052f803e4acb079bfdbefa9cc865af8b0d3f06a

crookedneighbor commented:
@JiriChara I've awarded you your first contributor tier! You can [read more about contributor awards here](http://habitrpg.wikia.com/wiki/Contributor_Rewards).
